tests/stub/authorization/test_temporal_auth_manager.py

```diff
@@ -0,0 +1,319 @@
+        self.post_script_assertions(self._reader)
+
+    @driver_feature(types.Feature.BACKEND_MOCK_TIME)
+    def test_static_provider_long_time(self):
+        count = 0
+
+        def provider():
+            nonlocal count
+            count += 1
+            return types.TemporalAuthToken(
+                types.AuthorizationToken(
+                    scheme="basic",
+                    principal="neo4j",
+                    credentials="pass"
+                )
+            )
+
+        auth_manager = TemporalAuthTokenManager(self._backend, provider)
+
+        self.start_server(
+            self._reader,
+            self.script_fn_with_features("reader_no_reauth.script")
+        )
+
+        with FakeTime(self._backend) as time:
+            with self.driver(auth_manager) as driver:
+                with self.session(driver) as session:
+                    list(session.run("RETURN 1 AS n"))
+                # just under 1 hour to make sure to not trip over the
+                # connection max lifetime
+                time.tick(1000 * 3600 - 1)
+                with self.session(driver) as session:
+                    # should still use the same token without calling the
+                    # provider
+                    list(session.run("RETURN 1 AS n"))
+
+        self._reader.done()
+        self.assertEqual(count, 1)
+        self.post_script_assertions(self._reader)
+
+    @driver_feature(types.Feature.BACKEND_MOCK_TIME)
+    def test_expiring_token_deadline(self):
+        count = 0
+
+        def provider():
+            nonlocal count
+            count += 1
+            credentials = "password" if count > 1 else "pass"
+
+            return types.TemporalAuthToken(
+                types.AuthorizationToken(
+                    scheme="basic",
+                    principal="neo4j",
+                    credentials=credentials
+                ),
+                10_000
+            )
+
+        auth_manager = TemporalAuthTokenManager(self._backend, provider)
+
+        self.start_server(self._reader,
+                          self.script_fn_with_features("reader_reauth.script"))
+        with FakeTime(self._backend) as time:
+            with self.driver(auth_manager) as driver:
+                with self.session(driver) as session:
+                    list(session.run("RETURN 1 AS n"))
+
+                time.tick(9_000)
+
+                with self.session(driver) as session:
+                    list(session.run("RETURN 2 AS n"))
+
+                self.assertEqual(1, count)
+
+                time.tick(1_001)
+
+                with self.session(driver) as session:
+                    list(session.run("RETURN 3 AS n"))
+
+                with self.session(driver) as session:
+                    list(session.run("RETURN 4 AS n"))
+
+                self.assertEqual(2, count)
+
+        self._reader.done()
+        self.post_script_assertions(self._reader)
+
+    def test_renewing_on_expiration_error(self):
```

Comment:
I think a different approach is needed for managing connections in this test. Java driver releases connections in background and lets user move on earlier. This means that the user is less likely to get the same connection in blocks like this:
```
# free connection 1
s1_tx.commit()
# bind connection 1
s1_tx = session_r1.begin_transaction()
```

Reply:
I think limiting the pool size to 3 and binding connection 2 after `# connection 2 fails, gets closed` so that the pool is again at its limit might be an option to work around this. Do you have capacity to try to make that work? We could also pair up on this.

Reply:
I have just pushed a several commits, please take a look when you can.

Reply:
I have just pushed several commits, please take a look when you can.